File: 04-structural-patterns/03-composite/03-composite.py
from abc import ABC, abstractmethod

class FileSystemComponent(ABC):
    @abstractmethod
    def show(self, indentation=""):
        pass

class File(FileSystemComponent):
    def __init__(self, name):
        self.name = name

    def show(self, indentation=""):
        print(f"{indentation}{self.name}")

class Folder(FileSystemComponent):
    def __init__(self, name):
        self.name = name
        self.children = []

    def add(self, component):
        self.children.append(component)

    def show(self, indentation=""):
        print(f"{indentation}{self.name}/")
        for child in self.children:
            child.show(indentation + "  ")

if __name__ == "__main__":
    root = Folder("root")
    home = Folder("home")
    user = Folder("user")
    file1 = File("file1.txt")
    file2 = File("file2.txt")

    user.add(file1)
    home.add(user)
    root.add(home)
    root.add(file2)

    root.show()

File: 04-structural-patterns/03-composite/03-composite.go
package main

import "fmt"

type FileSystemComponent interface {
	Show(indentation string)
}

type File struct {
	Name string
}

func (f *File) Show(indentation string) {
	fmt.Println(indentation + f.Name)
}

type Folder struct {
	Name     string
	Children []FileSystemComponent
}

func (f *Folder) Add(child FileSystemComponent) {
	f.Children = append(f.Children, child)
}

func (f *Folder) Show(indentation string) {
	fmt.Println(indentation + f.Name + "/")
	for _, child := range f.Children {
		child.Show(indentation + "  ")
	}
}

func main() {
	root := &Folder{Name: "root"}
	home := &Folder{Name: "home"}
	user := &Folder{Name: "user"}
	file1 := &File{Name: "file1.txt"}
	file2 := &File{Name: "file2.txt"}

	user.Add(file1)
	home.Add(user)
	root.Add(home)
	root.Add(file2)

	root.Show("")
}
